File: 0399-evaluate-division/0399-evaluate-division.java
import java.util.*;

class Solution {
    private HashMap<String, HashMap<String, Double>> graph = new HashMap<>();

    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        // Construct the graph from the equations
        for (int i = 0; i < equations.size(); i++) {
            String num = equations.get(i).get(0);
            String den = equations.get(i).get(1);
            double val = values[i];
            graph.computeIfAbsent(num, k -> new HashMap<>()).put(den, val);
            graph.computeIfAbsent(den, k -> new HashMap<>()).put(num, 1.0 / val);
        }

        // Handle the queries
        double[] result = new double[queries.size()];
        for (int i = 0; i < queries.size(); i++) {
            String num = queries.get(i).get(0);
            String den = queries.get(i).get(1);
            if (!graph.containsKey(num) || !graph.containsKey(den)) {
                result[i] = -1.0;
            } else {
                HashSet<String> visited = new HashSet<>();
                result[i] = dfs(num, den, 1.0, visited);
            }
        }

        return result;
    }

    private double dfs(String start, String end, double product, Set<String> visited) {
        // If we've visited the node before, return -1.0
        if (visited.contains(start)) {
            return -1.0;
        }
        // If we've reached our destination, return the product
        if (start.equals(end)) {
            return product;
        }
        visited.add(start);

        // Continue DFS on all neighbours
        for (String neighbor : graph.get(start).keySet()) {
            double result = dfs(neighbor, end, product * graph.get(start).get(neighbor), visited);
            if (result != -1.0) {
                return result;
            }
        }

        return -1.0;
    }
}

File: 0399-evaluate-division/0399-evaluate-division.kt
class Solution {
    private val graph = mutableMapOf<String, MutableMap<String, Double>>()

    fun calcEquation(equations: List<List<String>>, values: DoubleArray, queries: List<List<String>>): DoubleArray {
        for (i in equations.indices) {
            val equation = equations[i]
            graph.getOrPut(equation[0], { mutableMapOf() })[equation[1]] = values[i]
            graph.getOrPut(equation[1], { mutableMapOf() })[equation[0]] = 1.0 / values[i]
        }

        val result = DoubleArray(queries.size)
        for (i in queries.indices) {
            result[i] = dfs(queries[i][0], queries[i][1], HashSet())
        }
        return result
    }

    private fun dfs(start: String, end: String, visited: HashSet<String>): Double {
        if (!graph.containsKey(start)) return -1.0
        if (graph[start]!!.containsKey(end)) return graph[start]!![end]!!

        visited.add(start)
        for ((neighbor, value) in graph[start]!!) {
            if (visited.contains(neighbor)) continue
            val product = dfs(neighbor, end, visited)
            if (product != -1.0) return value * product
        }
        return -1.0
    }
}
